### PR TITLE
Centralize and document low-level OSSA utilities

### PR DESCRIPTION
**Pull request by atrick**
...for handling borrow scopes:

- find[Extended]TransitiveGuaranteedUses

- BorrowingOperand::visit[Extended]ScopeEndingUses

- BorrowedValue BorrowingOperand::getBorrowIntroducingUserResult()

And document the logic.

Mostly NFC in this commit, but more RAUW cases should be correctly
handled now.

Particularly, ensure that we can cleanly handle all manner of
reborrows. This is necessary to ensure both CanonicalizeOSSA and
replace-all-uses higher-level utilities handle all cases.

This generalizes some of the logic in CanonicalizeOSSA so it can be
shared by other high-level OSSA utilities.

These utilities extend the fundamental BorrowingOperand and
BorrowedValue functionality that has been designed recently. It builds
on and replaces a mix of piece-meal functionality that was needed
during bootstrapping. These APIs are now consistent with the more
recently designed code. It should be obvious what they mean and how to
use them, should be very hard to use them incorrectly, and should be
as efficient as possible, since they're fundamental to other
higher-level utilities.

Most importantly, there were several very subtle correctness issues
that were not handled cleanly in a centralized way. There are now a
mix of higher-level utilities trying to use this underlying
functionality, but it was hard to tell if all those higher-level
utilities were covering all the subtle cases:

- checking for PointerEscapes everywhere we need to

- the differences between uses of borrow introducers and other
  guaranteed values

- handling the uses of nested borrow scopes

- transitively handling reborrows

- the difference between nested and top-level reborrows

- forwarding destructures (which can cause exponential explosion)

In short, I was fundamentally confused and getting things wrong before
designing these utilities.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
Test cases are in the follow-up PRs that build on top of this

**Comment by atrick** [comment]
There seems to be something wrong with an llbuild/Ninja test
```Failure: Ninja/Build/max-commands-progress.ninja

max-commands-progress.ninja:10:18: error: CHECK-INITIAL: expected string not found in input
23:19:46 # CHECK-INITIAL: [1/{{.*}}] cp input-2 output-2
23:19:46                  ^

note: possible intended match here
23:19:46 [2/3] cp input-2 output-2
23:19:46 ^
```

**Comment by atrick** [comment]
@swift-ci smoke test macOS

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 1]
```diff
@@ -878,6 +845,160 @@ OwnedValueIntroducer swift::getSingleOwnedValueIntroducer(SILValue inputValue) {
   llvm_unreachable("Should never hit this");
 }
 
+// Find all use points of \p guaranteedValue within its borrow scope where
+// \p guaranteedValue is not itself a borrow introducer.
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:849]: Rather than use Borrow Introducer, I would prefer if we centralized on BorrowedValue and talk about those introducing the borrow. Consistency I think is going to be important. We should talk over the comments tomorrow.

**Review by gottesmm — COMMENTED** [review]
Some initial comments. I need to read through it tomorrow locally in the editor.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 2]
```diff
@@ -244,14 +162,39 @@ static bool canFixUpOwnershipForRAUW(SILValue oldValue, SILValue newValue,
...
+    auto *branch = endScope->getUser();
+    context.recursiveReborrows.push_back(
+      {branch->getParent(), endScope->getOperandNumber()});
+    return true;
+  };
+  if (!findTransitiveGuaranteedUses(oldValue, context.transitiveBorrowedUses,
+                                    visitReborrow))
+    return false;
+
+  for (unsigned idx = 0; idx < reborrows.size(); ++idx) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:190]: Can you add a comment here saying that reborrows grows in the loop.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 3]
```diff
@@ -244,14 +162,39 @@ static bool canFixUpOwnershipForRAUW(SILValue oldValue, SILValue newValue,
...
-    }
+  if (oldValue.getOwnershipKind() != OwnershipKind::Guaranteed)
+    return true;
+
+  // Check that the old lifetime can be extended and record the necessary
+  // book-keeping in the OwnershipFixupContext.
+  context.clear();
+
+  // FIXME: the rest of the logic in this function a shim between OwnershipUtils
+  // and OwnershipContextFixup that will be simplified in a follow-up commit.
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:173]: Can you put in a comment here something that explains that we are implementing the same thing as the extended lifetime version but we want to have more control. Even if this is going away, when I was reading the code, the way it is laid out made me look for bugs there. So a more explicit comment connecting the two together would be appreciated.

**Review by gottesmm — APPROVED** [review]
A little bit more feedback. Lets iterate on the larger comments later on subsequent PRs.

**Comment by atrick** [comment]
@gottesmm The documentation review converged to this:
```
/// Find all "use points" of a guaranteed value within its enclosing borrow
/// scope (without looking through reborrows). To find the use points of the
/// extended borrow scope, after looking through reborrows, use
/// findExtendedTransitiveGuaranteedUses() instead.
///
/// Accumulate results in \p usePoints. This avoids the need for separate
/// worklist and result vectors. Existing vector elements are ignored.
///
/// "Use points" are the relevant points for determining lifetime. They are
/// determined differently depending on each of these two cases:
///
/// 1. If \p guaranteedValue introduces a borrow scope (begin_borrow,
/// load_borrow, or phi), then its only use points are the scope-ending uses,
/// and this function returns true. This is, in fact, equivalent to calling
/// BorrowedValue::visitLocalScopeEndingUses(). Any scope-ending uses that are
/// reborrows are recorded as use points without following the reborrowed
/// uses. The \p visitReborrow callback can be used to transitively process
/// reborrows to discover the extended lifetime. Reborrows may be recursive, so
/// this will require checking membership in a working set. Nested borrow scope
/// are irrelevant to the parent scope's lifetime. They are not considered use
/// points, and reborrows within those nested scope are not visited by \p
/// visitReborrow.
///
/// 2. If \p guaranteedValue does not introduce a borrow scope (it is not a
/// valid BorrowedValue), then its uses are discovered transitively by looking
/// through forwarding operations. If any use is a PointerEscape, then this
/// returns false without adding more uses--the guaranteed values lifetime is
/// indeterminite. If a use introduces a nested borrow scope, it creates use
/// points where the "extended" borrow scope ends. An extended borrow
/// scope is found by looking through any reborrows that end the nested
/// scope. Other uses within nested borrow scopes are ignored.
bool findTransitiveGuaranteedUses(SILValue guaranteedValue,
                                  SmallVectorImpl<Operand *> &usePoints,
                                  function_ref<bool(Operand *)> visitReborrow);

/// Find all "use points" of guaranteed value across its extended borrow scope
/// (looking through reborrows). The "use points" are the relevant points for
/// determining lifetime.
///
/// Accumulate results in \p usePoints. This avoids the need for separate
/// worklist and result vectors. Existing vector elements are ignored.
///
/// "Use points" are the relevant points for determining lifetime. They are
/// determined differently depending on each of these two cases:
///
/// 1. If \p guaranteedValue introduces a borrow scope (begin_borrow,
/// load_borrow, or phi), then its only use points are the extended scope-ending
/// uses, and this function returns true. This is, in fact, equivalent to
/// calling BorrowedValue::visitExtendedLocalScopeEndingUses().
///
/// 2. If \p guaranteedValue does not introduce a borrow scope (it is not a
/// valid BorrowedValue), then its uses are discovered transitively by looking
/// through forwarding operations. Only a BorrowedValue can have its lifetime
/// extended by a reborrow; therefore, in this case, the algorithm is equivalent
/// to findTransitiveGuaranteedUses(). See those comments for more detail.
bool findExtendedTransitiveGuaranteedUses(
  SILValue guaranteedValue,
  SmallVectorImpl<Operand *> &usePoints);
```

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci smoke test
